metrics/metric_test.go

```diff
@@ -66,3 +66,104 @@ func TestAddSubmetric(t *testing.T) {
 		})
 	}
 }
+
+func TestParseMetricName(t *testing.T) {
+	// TODO: verify we get an ErrMetricNameParsing
+	t.Parallel()
+
+	tests := []struct {
+		name                 string
+		metricNameExpression string
+		wantMetricName       string
+		wantTags             []string
+		wantErr              bool
+	}{
+		{
+			name:                 "metric name without tags",
+			metricNameExpression: "test_metric",
+			wantMetricName:       "test_metric",
+			wantTags:             nil,
+			wantErr:              false,
+		},
+		{
+			name:                 "metric name with single tag",
+			metricNameExpression: "test_metric{abc:123}",
+			wantMetricName:       "test_metric",
+			wantTags:             []string{"abc:123"},
+			wantErr:              false,
+		},
+		{
+			name:                 "metric name with multiple tags",
+			metricNameExpression: "test_metric{abc:123,easyas:doremi}",
+			wantMetricName:       "test_metric",
+			wantTags:             []string{"abc:123", "easyas:doremi"},
+			wantErr:              false,
+		},
+		{
+			name:                 "metric name with multiple spaced tags",
+			metricNameExpression: "test_metric{abc:123, easyas:doremi}",
+			wantMetricName:       "test_metric",
+			wantTags:             []string{"abc:123", "easyas:doremi"},
+			wantErr:              false,
+		},
+		{
+			name:                 "metric name with group tag",
+			metricNameExpression: "test_metric{group:::mygroup}",
+			wantMetricName:       "test_metric",
+			wantTags:             []string{"group:::mygroup"},
+			wantErr:              false,
+		},
+		{
+			name:                 "metric name with tag definition missing `:value`",
+			metricNameExpression: "test_metric{easyas}",
+			wantMetricName:       "",
+			wantTags:             nil,
+			wantErr:              true,
+		},
+		{
+			name:                 "metric name with tag definition missing value",
+			metricNameExpression: "test_metric{easyas:}",
+			wantMetricName:       "",
+			wantTags:             nil,
+			wantErr:              true,
+		},
+		{
+			name:                 "metric name with mixed valid and invalid tag definitions",
+			metricNameExpression: "test_metric{abc:123,easyas:}",
+			wantMetricName:       "",
+			wantTags:             nil,
+			wantErr:              true,
+		},
```

Comment:
Can you add a bunch more cases with even more misformed metrics such as `test_metric{` and such

https://github.com/grafana/k6/pull/2426#discussion_r834502077 has some other examples.

Also, you can skip setting `wantMetricName`, `withTags`, `wantErr`, when it's their "nil" values - it will save some space and I don't think it makes it any less readable - I would argue it does the opposite

Reply:
I think you've addressed this while I went through the rest of the code

Reply:
I've removed the `wantMetric`, `withTags` and `wantErr` where relevant.
I've also added the test case for https://github.com/grafana/k6/pull/2426#discussion_r834502077.